nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/StreamBulkScroller.java

```diff
@@ -0,0 +1,173 @@
+        }
+        return Topology.builder()
+                       .addComputation( //
+                               () -> new BulkDocumentScrollerComputation(COMPUTATION_NAME, operations.size(), kvStore,
+                                       scrollBatchSize, scrollKeepAliveSeconds), //
+                               mapping)
+                       .build();
+    }
+
+    public static class BulkDocumentScrollerComputation extends AbstractComputation {
+
+        protected final KeyValueStore kvStore;
+
+        protected final int scrollBatchSize;
+
+        protected final int scrollKeepAliveSeconds;
+
+        public BulkDocumentScrollerComputation(String name, int nbOutputStreams, String kvStore, int scrollBatchSize,
+                int scrollKeepAliveSeconds) {
+            super(name, 1, nbOutputStreams);
+            this.kvStore = Framework.getService(KeyValueService.class).getKeyValueStore(kvStore);
+            this.scrollBatchSize = scrollBatchSize;
+            this.scrollKeepAliveSeconds = scrollKeepAliveSeconds;
+        }
+
+        @Override
+        public void processRecord(ComputationContext context, String inputStreamName, Record record) {
+            TransactionHelper.runInTransaction(() -> processRecord(context, record));
+        }
+
+        protected void processRecord(ComputationContext context, Record record) {
+            try {
+                BulkCommand command = getBulkCommandJson(record.data);
+                if (!kvStore.compareAndSet(record.key + STATE, SCHEDULED.toString(), BUILDING.toString())) {
+                    log.error("Discard record: " + record + " because it's already building");
+                    return;
+                }
+                try (CloseableCoreSession session = CoreInstance.openCoreSession(command.getRepository(),
+                        command.getUsername())) {
+                    // scroll documents
+                    ScrollResult<String> scroll = session.scroll(command.getQuery(), scrollBatchSize,
+                            scrollKeepAliveSeconds);
+                    long documentCount = 0;
+                    while (scroll.hasResults()) {
+                        List<String> docIds = scroll.getResults();
+                        // send these ids as keys to the appropriate stream
+                        // key will be bulkOperationId/docId
+                        // value/data is a BulkCommand serialized as JSON
+                        docIds.forEach(docId -> context.produceRecord(command.getOperation(), record.key + '/' + docId,
+                                record.data));
+                        documentCount += docIds.size();
+                        // next batch
+                        scroll = session.scroll(scroll.getScrollId());
+                        TransactionHelper.commitOrRollbackTransaction();
+                        TransactionHelper.startTransaction();
+                    }
+                    context.askForCheckpoint();
```

Comment:
records are send only on checkpoint, this means that all the record are accumulated until here which is not going to work for millions of records. For this special computation which can generates lots of records it might be ok to askForCheckpoint for each scroll results.